### PR TITLE
Add basis set functions for HF solver

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
This PR adds a class and functions needed to create a basis function object and generate default basis set parameters. The basis function object holds the basis set parameters and is used as input to compute various integrals over basis functions, such as the overlap integral, that are needed to perform a Hartree-Fock calculation.

**Description of the Change:**
The BasisFunction class is added to creates basis function objects and the functions atom_basis_data and mol_basis_data are added to generate the default basis set parameters. The parameters of the STO-3G basis set, including the exponents and contraction coefficients of the Gaussian functions and the angular momentum numbers, are stored in a dictionary for atoms with atomic numbers 1-10. The parameter set can be easily extended to contain more atoms and more basis sets.

**Benefits:**
The BasisFunction object makes the basis set parameters easy to handle and manipulate. Having the default basis set parameters makes PL independent of external libraries, such as the Basis Set Exchange [library](https://www.basissetexchange.org/), for generating basis set data. Note that the users still have the option of using their own basis set data imported from any desired library in case they prefer not to use the default parameters provided in this PR. 

**Possible Drawbacks:**
No specific drawbacks.

**Related GitHub Issues:**
No related issues.

### CODE REVIEW

File: qchem/pennylane_qchem/hf/basis_set.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        name (str): name of the basis set
+        symbols (list[str]): symbols of the atomic species in the molecule
+
+    Returns:
+        tuple(list, tuple): the number of basis functions and the basis set parameters for each atom
+        in the molecule
+
+    **Example**
+
+    >>>  n_basis, params = mol_basis_data('sto-3g', ['H', 'H'])
```
Comment (rmoyard) [inline qchem/pennylane_qchem/hf/basis_set.py:123]:
```suggestion
    >>> n_basis, params = mol_basis_data('sto-3g', ['H', 'H'])
```

**Review by rmoyard — COMMENTED** [review]
Nice PR, my main comment would be to create an `__init__.py` in both `pennylane_qchem` and `hf`, then it is possible to use `hf` function outside this scope. By doing this you can also move the `test_basis_set.py`file to the main `tests`folder. I have something almost ready if you agree with that. You can also add the change to the qchem `CHANGELOG.md`.

File: qchem/pennylane_qchem/hf/basis_set.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        self.alpha = alpha
+        self.coeff = coeff
+        self.rgaus = rgaus
+        self.params = [self.alpha, self.coeff, self.rgaus]
+
+
+def atom_basis_data(name, atom):
+    r"""Generate default basis set parameters for an atom.
+
+    This function extracts the default angular momentum and the exponents and contraction
```
Comment (agran2018) [inline qchem/pennylane_qchem/hf/basis_set.py:69]: What do you mean by `default` angular momentum?

File: qchem/pennylane_qchem/hf/basis_set.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    **Example**
+
+    >>>  params = atom_basis_data('sto-3g', 'H')
+    >>> print(params)
+    [((0, 0, 0), [3.425250914, 0.6239137298, 0.168855404], [0.1543289673, 0.5353281423, 0.4446345422])]
+    """
+    basis_sets = {"sto-3g": STO3G}
+
+    s = [(0, 0, 0)]
+    p = [(0, 0, 1), (0, 1, 0), (1, 0, 0)]
```
Comment (agran2018) [inline qchem/pennylane_qchem/hf/basis_set.py:90]: is the convention here px = (1, 0, 0), py = (0, 1, 0), pz = (0, 0, 1) ?

File: qchem/pennylane_qchem/hf/basis_set.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        if j == "S":
+            params.append((s[0], basis["exponents"][i], basis["coefficients"][i]))
+        elif j == "SP":
+            for term in j:
+                if term == "S":
+                    params.append((s[0], basis["exponents"][i], basis["coefficients"][i]))
+                if term == "P":
+                    for l in p:
+                        params.append((l, basis["exponents"][i], basis["coefficients"][i + 1]))
+    return params
```
Comment (agran2018) [inline qchem/pennylane_qchem/hf/basis_set.py:104]: Nice!

File: qchem/pennylane_qchem/hf/basis_set.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+    This function generates the default basis set parameters for a list of atomic symbols and
+    computes the total number of basis functions for each atom.
+
+    Args:
+        name (str): name of the basis set
+        symbols (list[str]): symbols of the atomic species in the molecule
+
+    Returns:
+        tuple(list, tuple): the number of basis functions and the basis set parameters for each atom
```
Comment (agran2018) [inline qchem/pennylane_qchem/hf/basis_set.py:118]:
```suggestion
        tuple(list, tuple): the number of atomic basis functions and the basis set parameters for each atom
```

File: qchem/pennylane_qchem/hf/basis_set.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+    **Example**
+
+    >>>  n_basis, params = mol_basis_data('sto-3g', ['H', 'H'])
+    >>> print(n_basis)
+    [1, 1]
+    >>> print(params)
+    (((0, 0, 0), [3.425250914, 0.6239137298, 0.168855404], [0.1543289673, 0.5353281423, 0.4446345422]),
+     ((0, 0, 0), [3.425250914, 0.6239137298, 0.168855404], [0.1543289673, 0.5353281423, 0.4446345422]))
+    """
```
Comment (agran2018) [inline qchem/pennylane_qchem/hf/basis_set.py:129]: May we use for the example a molecule with two different atoms e.g., LiH?  A less symmetric output might be more clarifying.

**Review by agran2018 — CHANGES_REQUESTED** [review]
Thank you @soranjh! I have left some suggestions. Very interesting PR.

**Comment by soranjh** [comment]
> 2. Have a _different_ module name for the built-in qchem components of PennyLane.

@josh146, @rmoyard, I have created the qchem_hf module in pennylane to store the differentiable HF functions. Any thoughts?

**Comment by josh146** [comment]
> I have created the qchem_hf module in pennylane to store the differentiable HF functions. Any thoughts?

Nice 👍 Only suggestion I have is minor, but `pennylane.hf` is quite concise and nice sounding to my ears!

Don't forget to move the changelog entry to the PL changelog :)

**Comment by soranjh** [comment]
> > I have created the qchem_hf module in pennylane to store the differentiable HF functions. Any thoughts?
> 
> Nice  Only suggestion I have is minor, but `pennylane.hf` is quite concise and nice sounding to my ears!
> 
> Don't forget to move the changelog entry to the PL changelog :)

Agree. Thanks Josh.

**Comment by rmoyard** [comment]
@soranjh Yes I think it is a good solution and agree with Josh !

File: pennylane/hf/basis_set.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains functions and classes to create a BasisFunction object from standard basis sets
+such as STO-3G.
+"""
+# pylint: disable=too-few-public-methods
+from pennylane.qchem_hf.basis_data import STO3G
```
Comment (rmoyard) [inline pennylane/hf/basis_set.py:19]:
```suggestion
from pennylane.hf.basis_data import STO3G
```

File: tests/hf/test_basis_set.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Unit tests for generating basis set default parameters.
+"""
+# pylint: disable=no-self-use
+import pytest
+from pennylane import numpy as np
+from pennylane.qchem_hf.basis_set import BasisFunction, atom_basis_data, mol_basis_data
```
Comment (rmoyard) [inline tests/hf/test_basis_set.py:20]:
```suggestion
from pennylane.hf.basis_set import BasisFunction, atom_basis_data, mol_basis_data
```

**Review by rmoyard — COMMENTED** [review]
You just forgot to change the name in the import 👍

**Pull request merged**